File: sinteemar/models/usuario.py
from datetime import datetime
from hashlib import md5
from re import search
from sinteemar.models.permissao import Permissao

class Usuario():
	'''
	Atributos:
		id (int): Chave primária
		nome (str): Nome completo
		email (str): Endereço de e-mail
		login (str): Apelido único
		senha (str): Senha alfanumérica
		permissao (Permissao): Permissões
		data (datetime): Data de cadastro do usuário
		status (bool): Inativo (False) ou ativo (True)
	'''
	__slots__ = ['_ativo', '_data', '_email', '_id', '_login', '_nome', '_permissao', '_senha']

	def __init__(self, id: int=0, nome: str='', email: str='', login: str='', senha: str='', permissao: Permissao=Permissao(), data: datetime=datetime.now(), ativo: bool=False):
		self._id = id
		self._nome = nome
		self._email = email
		self._login = login
		self._senha = senha
		self._permissao = permissao
		self._data = data
		self._ativo = ativo

	def __str__(self) -> str:
		return 'ID: ' + str(self._id) + '\nNOME: ' + self._nome.title() + '\nE-MAIL: ' + self._email + '\nLOGIN: ' + self._login + '\nSENHA: ' + self._senha + '\nPERMISSÃO: ' + str(self._permissao.id) + '\nDATA: ' + self._data.strftime('%d/%m/%Y - %H:%M:%S') + '\nATIVO: ' + str(self._ativo)

	@property
	def id(self) -> int:
		return self._id

	@id.setter
	def id(self, id: int) -> bool:
		if id > 0:
			self._id = id
			return True
		return False

	@property
	def nome(self) -> str:
		return self._nome

	@nome.setter
	def nome(self, nome: str) -> bool:
		nome = nome.strip()
		if len(nome) in range(6, 65):
			self._nome = nome
			return True
		return False

	@property
	def email(self) -> str:
		return self._email

	@email.setter
	def email(self, email: str) -> bool:
		email = email.strip()
		if len(email) in range(6, 65) and search(r'(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)', email):
			self._email = email.lower()
			return True
		return False

	@property
	def login(self) -> str:
		return self._login

	@login.setter
	def login(self, login: str) -> bool:
		login = login.strip()
		if len(login) in range(6, 65):
			self._login = login.lower()
			return True
		return False

	@property
	def senha(self) -> str:
		return self._senha

	@senha.setter
	def senha(self, senha: str|tuple[str, bool]) -> bool:
		criptografa = False
		try:
			senha, criptografa = senha
		except ValueError:
			pass
		if len(senha) >= 6:
			if criptografa:
				self._senha = md5(str.encode(senha)).hexdigest()
			else:
				self._senha = senha
			return True
		return False

	@property
	def permissao(self) -> Permissao:
		return self._permissao

	@permissao.setter
	def permissao(self, permissao: Permissao) -> bool:
		self._permissao = permissao
		return True

	@property
	def data(self) -> datetime:
		return self._data

	@data.setter
	def data(self, data: str) -> bool:
		try:
			self._data = datetime(int(data[:4]), int(data[5:7]), int(data[8:10]), int(data[11:13]), int(data[14:16]), int(data[17:19]))
			return True
		except TypeError:
			return False

	@property
	def ativo(self) -> bool:
		return self._ativo

	@ativo.setter
	def ativo(self, ativo: bool) -> bool:
		self._ativo = bool(ativo)
		return True

File: modelos/Usuario.php
<?php
	include_once("Permissao.php");
	/**
	 * Classe Usuário com atributos simples (ID, nome, e-mail, login, senha, permissão e status)
	 * @author Luiz Joaquim Aderaldo Amichi <luizamichi@luizamichi.com.br>
	 * @copyright 2020 Luiz Joaquim Aderaldo Amichi
	 */
	class Usuario {
		/**
		 * Variáveis privadas
		 * @var int $id Chave primária
		 * @var string $nome Nome completo
		 * @var string $email Endereço de e-mail
		 * @var string $login Apelido único
		 * @var string $senha Senha alfanumérica
		 * @var Permissao $permissao Permissão (ID, CRUD(banner), CRUD(boletim), CRUD(convenção), CRUD(convênio), LEITURA(diretório), CRUD(edital), CRUD(evento), CRUD(finança), CRUD(jornal), CRUD(jurídico), CRUD(notícia), CRUD(podcast), LEITURA(registro), LEITURA(tabela) e CRUD(usuário))
		 * @var boolean $status Inativo (FALSE) ou ativo (TRUE)
		 */
		private $id;
		private $nome;
		private $email;
		private $login;
		private $senha;
		private $permissao;
		private $status;

		/**
		 * Instancia a classe
		 * @return void
		 */
		public function __construct() {
			$this->id = 0;
			$this->nome = "";
			$this->email = "";
			$this->login = "";
			$this->senha = "";
			$this->permissao = new Permissao();
			$this->status = FALSE;
		}

		/**
		 * Destrói a classe
		 * @return void
		 */
		public function __destruct() {
			$this->id = NULL;
			$this->nome = NULL;
			$this->email = NULL;
			$this->login = NULL;
			$this->senha = NULL;
			$this->permissao = NULL;
			$this->status = NULL;
		}

		/**
		 * Clona a classe
		 * @return void
		 */
		public function __clone() {
			$this->id = clone $this->id;
			$this->nome = clone $this->nome;
			$this->email = clone $this->email;
			$this->login = clone $this->login;
			$this->senha = clone $this->senha;
			$this->permissao = clone $this->permissao;
			$this->status = clone $this->status;
		}

		/**
		 * Retorna a classe em formato de string
		 * @return string Atributos da classe separados por quebra de linha
		 */
		public function __toString() {
			return "ID: " . $this->id .
			"\nNOME: " . $this->nome .
			"\nE-MAIL: " . $this->email .
			"\nLOGIN: " . $this->login .
			"\nSENHA: " . $this->senha .
			"\nPERMISSÃO: " . $this->permissao->getId() .
			"\nSTATUS: " . ($this->status ? "ATIVO" : "INATIVO") . "\n";
		}

		/**
		 * Atribui um novo valor ao ID
		 * @param int $id Número inteiro positivo
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setId(int $id) {
			if($id > 0) {
				$this->id = $id;
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor ao nome
		 * @param string $nome Nome completo do usuário com no mínimo 6 e no máximo 64 caracteres
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setNome(string $nome) {
			$nome = trim($nome);
			if(strlen($nome) >= 6 && strlen($nome) <= 64) {
				$this->nome = mb_strtoupper($nome);
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor ao e-mail
		 * @param string $email E-mail no formato nome@dominio.com.br com no mínimo 6 e no máximo 64 caracteres
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setEmail(string $email) {
			$email = trim($email);
			if(filter_var($email, FILTER_VALIDATE_EMAIL) && strlen($email) >= 6 && strlen($email) <= 64) {
				$this->email = mb_strtolower($email);
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor ao login
		 * @param string $login Texto alfanumérico único vinculado ao usuário com no mínimo 6 e no máximo 64 caracteres
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setLogin(string $login) {
			$login = trim($login);
			if(strlen($login) >= 6 && strlen($login) <= 64) {
				$this->login = $login;
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor à senha
		 * @param string $senha Texto alfanumérico secreto vinculado ao usuário com no mínimo 6 caracteres
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setSenha(string $senha) {
			if(strlen($senha) >= 6) {
				$this->senha = $senha;
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor criptografado com MD5 à senha
		 * @param string $senha Texto alfanumérico secreto vinculado ao usuário com no mínimo 6 caracteres
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function hashSenha(string $senha) {
			if(strlen($senha) >= 6) {
				$this->senha = md5($senha);
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor em branco à senha
		 * @return void
		 */
		public function limpaSenha() {
			$this->senha = "";
		}

		/**
		 * Atribui uma permissão (objeto) ao usuário
		 * @param Permissao $permissao Objeto instanciado
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function setPermissao(Permissao $permissao) {
			if($permissao->valida()) {
				$this->permissao = $permissao;
				return TRUE;
			}
			else {
				return FALSE;
			}
		}

		/**
		 * Atribui um novo valor ao status do usuário
		 * @param boolean $status Booleano referente ao estado
		 * @return void
		 */
		public function setStatus(bool $status) {
			$this->status = $status;
		}

		/**
		 * Retorna o ID do usuário
		 * @return int Inteiro positivo único
		 */
		public function getId() {
			return $this->id;
		}

		/**
		 * Retorna o nome do usuário
		 * @return string Nome completo
		 */
		public function getNome() {
			return $this->nome;
		}

		/**
		 * Retorna o e-mail do usuário
		 * @return string E-mail da caixa postal
		 */
		public function getEmail() {
			return $this->email;
		}

		/**
		 * Retorna o login do usuário
		 * @return string Chave única
		 */
		public function getLogin() {
			return $this->login;
		}

		/**
		 * Retorna a senha do usuário
		 * @return string Chave secreta
		 */
		public function getSenha() {
			return $this->senha;
		}

		/**
		 * Retorna as permissões do usuário
		 * @return Permissao Objeto permissão (ID, CRUD(banner), CRUD(boletim), CRUD(convenção), CRUD(convênio), LEITURA(diretório), CRUD(edital), CRUD(evento), CRUD(finança), CRUD(jornal), CRUD(jurídico), CRUD(notícia), CRUD(podcast), LEITURA(registro), LEITURA(tabela) e CRUD(usuário))
		 */
		public function getPermissao() {
			return $this->permissao;
		}

		/**
		 * Retorna o status de atividade do usuário
		 * @return boolean Inativo ou ativo
		 */
		public function isStatus() {
			return $this->status;
		}

		/**
		 * Valida o usuário certificando todos os atributos
		 * @return boolean Sucesso ou fracasso na operação
		 */
		public function valida() {
			if($this->id && $this->nome && $this->email && $this->login && $this->senha && $this->permissao->valida() && $this->status) {
				return TRUE;
			}
			else {
				return FALSE;
			}
		}
	}
?>